app/javascript/utilities/viewport.js

```diff
@@ -3,22 +3,45 @@
  *
  * @example
  * const element = document.getElementById('element');
- * isInViewport(element); // true or false
+ * isInViewport({element, allowPartialVisibility = true}); // true or false
  *
  * @param {object} element - The HTML element to check
  * @param {number} [offsetTop=0] - Part of the screen to ignore counting from the top
- *
+ * @param {boolean} [allowPartialVisibility=false] - A boolean to flip the check between partial or completely visible in the viewport
  * @returns {boolean} isInViewport - true if the element is visible in the viewport
  */
-export function isInViewport(element, offsetTop = 0) {
+export function isInViewport({
+  element,
+  offsetTop = 0,
+  allowPartialVisibility = false,
+}) {
   const boundingRect = element.getBoundingClientRect();
   const clientHeight =
     window.innerHeight || document.documentElement.clientHeight;
   const clientWidth = window.innerWidth || document.documentElement.clientWidth;
+  const topIsInViewport =
+    boundingRect.top <= clientHeight && boundingRect.top >= offsetTop;
+  const rightIsInViewport =
```

Comment:
question (potentially blocking): Is it necessary to check if the right and left of the element is in the viewport? Shouldn't top and bottom be enough?

Also, since this is being revisited, we could probably leverage the [Intersection Observer API](https://developer.mozilla.org/en-US/docs/Web/API/Intersection_Observer_API)? Or maybe save that for a subsequent PR, cc: @aitchiss.

Reply:
Yeah I agree it would be nice to rework this to use intersection observer - I'd be happy to tackle in a later PR though, since the current implementation fixes the bug 👍

Reply:
Hey, @nickytonline! I have avoided using the Intersection observer since I felt it will be a little overhead. Intersection Observer will be listening to the viewport and callbacks will be triggered when articles come in and leave the viewport. I felt that is not what is required here. What we need is when 'j' and 'k' key events happen, we have to let the user select some element based on what is visible in the viewport. I think the requirement here is to select from what is present in the viewport rather than observing what comes in and leaves the viewport. So felt the Intersection API might not be necessary. Correct me if my understanding is wrong.

Reply:
And also regarding the _left_ and _right_ check. Since this "_isInViewport_" function is part of the utilities I had included the left and right check as well to fit its correct definition. If it is specific for only this case I can remove those checks as well.